Add GitHub workflow documentation (#256)
WordPress/performance: docs/Github-workflow.md

> [Back to overview](https://github.com/WordPress/performance/blob/trunk/docs/README.md)

Reviewer: Nit-pick, but it's better to have this a relative URL so that it works no matter which branch you're looking at.
```suggestion
[Back to overview](./README.md)
```

# GitHub workflow for the Performance Lab plugin

> The Performance team uses GitHub to manage all code and related discussions for the Performance Lab plugin. Please follow the workflow below to ensure that all issues are properly tracked.

Reviewer 2: The "Performance team" could have an external link to https://make.wordpress.org/core/2021/10/12/proposal-for-a-performance-team/

## Issues
When [opening a new issue](https://github.com/WordPress/performance/issues/new/choose), use the appropriate template: Bug report, Feature request, or Report a security vulnerability. All new issues should include the following labels:

- A _[Focus]_ label, or Infrastructure if the issue relates to the plugin infrastructure. The [Focus] labels are aligned with the Performance team's [focus areas](https://docs.google.com/spreadsheets/d/16N5oZ9wE6AkiqMz7b_707eh24vvpjMwsEG67XFAbxy8/edit#gid=0) and [GitHub Projects](https://github.com/WordPress/performance/projects).
- A _[Type]_ label
- A _[Module]_ label if the issue relates to an existing module

In addition, the new issue should be assigned to an appropriate Project. By default, new issues will automatically be added to the Backlog column within its project. This is intended for tracking any future work that is not currently a high priority as defined by the project’s point(s) of contact (POCs). Contributors are welcome to work on issues in the Backlog, but response rates may be slower than they are on prioritized issues.

> ## Working on an issue

Reviewer: I think this belongs under the "Issues" heading or not? In other words, I'd make this a `h3`, and then you could make all the labels below `h4`.

If you’re interested in working on an issue, it's helpful to  notify the POC for the issue’s focus area by tagging them on the issue and/or notifying them in the [weekly performance chat](https://make.wordpress.org/core/tag/performance/) so that everyone is aware that someone is working on it and effort is not duplicated. For an updated list of POCs, please [see the `CODEOWNERS` file](https://github.com/WordPress/performance/blob/trunk/.github/CODEOWNERS).

When you’re ready to begin work on an issue:

> - Assign it to yourself – Depending on your permissions levels, you may not be able to self-assign an issue. Please tag @bethanylang on an issue if you need assistance with permissions. Please only assign an issue if you plan to work on it within a reasonable time frame, i.e. within the next 2 weeks. 

Reviewer 2: I think it would be better to tag @WordPress/performance-admins to avoid relying on individuals

- Change the Project column to **In Progress**
- Remove the **Needs Dev** label, if applicable

In addition, there are several **Needs** labels that can be used to clarify next steps on an issue.

### Needs Discussion
Many issues require discussion and definition before development begins. For example, multiple approaches may be considered and the community should discuss them before a contributor proceeds with development. For those issues, add the **Needs Discussion** label and raise the issue in the weekly performance chat.

### Needs Decision

> After discussion, a formal vote may be needed to determine how to proceed. If that’s the case, please tag @bethanylang in the issue for assistance with setting up a vote via GitHub comment. [An example vote can be found here](https://github.com/WordPress/performance/issues/92#issuecomment-1068215411).

Reviewer 2: Same as the above, let tagging @WordPress/performance-admins is more appropriate

### Needs Dev
If an issue requires (more) development and you are unable to complete it yourself, remove yourself as the assignee and add a **Needs Dev** label.

### Needs Testing

> If you have completed initial engineering and want community members to test prior to merge, add the Needs Testing label.

Reviewer: See above, I think these headings here would be better as `h4`, under "Working on an issue" which would be `h3` based on my other comment.

> The Performance team uses GitHub to manage all code and related discussions for the Performance Lab plugin. Please follow the workflow below to ensure that all issues are properly tracked.

Reviewer: Can you add a `#` heading to the top of this doc, and also a link back to the overview, like it's in the other docs files?

Also, it would be good if you could add a link to this doc to the `README.md` overview file which acts as docs table of contents basically.

> ## Opening an issue

Reviewer: Just an idea: You have below other sections of the same type called "Pull requests" and then "Trac tickets". How about we make the `h2` here a single "Issues" headline and then the "Opening an issue" right below it, which would bring it to the same heading level as "Working on an issue"?

Here's what I mean:
```suggestion
## Issues

### Opening an issue
```

This way the document outline would become:
* GitHub workflow
    * Issues
    * Pull requests
    * Trac tickets
    * POC responsibilities

> When opening a new issue, use the appropriate template: Bug report, Feature request, or Report a security vulnerability. All new issues should include the following labels:

Reviewer 3: ```suggestion
When [opening a new issue](https://github.com/WordPress/performance/issues/new/choose), use the appropriate template: Bug report, Feature request, or Report a security vulnerability. All new issues should include the following labels:
```

> - A [Focus] label, or Infrastructure if the issue relates to the plugin infrastructure. The [Focus] labels are aligned with the Performance team's [focus areas](https://docs.google.com/spreadsheets/d/16N5oZ9wE6AkiqMz7b_707eh24vvpjMwsEG67XFAbxy8/edit#gid=0) and [GitHub Projects](https://github.com/WordPress/performance/projects).

Reviewer: A general note on formatting of these labels, it may be good to highlight them in some way. Maybe in italic (_[Focus]_) or code style (`[Focus]`)? I'm just leaving this comment here, but applies to any labels mentioned throughout this doc.

- A [Type] label
- A [Module] label if the issue relates to an existing module

> In addition, the new issue should be assigned to an appropriate Project: Object Cache, Measurement, Site Health, JavaScript, Infrastructure, or Images.

Reviewer: I'd prefer if we could avoid mentioning the projects that are available. Otherwise we'd have to remember modifying this doc every time we add one.

> In addition, the new issue should be assigned to an appropriate Project: Object Cache, Measurement, Site Health, JavaScript, Infrastructure, or Images.

Reviewer 3: ```suggestion
In addition, the new issue should be assigned to an appropriate Project: 

- Object Cache
- Measurement 
- Site Health
- JavaScript
- Infrastructure
- Images.
```

Reviewer: See my comment in https://github.com/WordPress/performance/pull/256#discussion_r833521448, I think it would be better to avoid mentioning the available projects altogether for maintenance reasons.

By default, new issues will automatically be added to the Backlog column within its project. This is intended for tracking any future work that is not currently a high priority as defined by the project’s point(s) of contact (POCs). Contributors are welcome to work on issues in the Backlog, but response rates may be slower than they are on prioritized issues.

### Working on an issue

> If you’re interested in working on an issue, please notify the POC for the issue’s focus area by tagging them on the issue and/or notifying them in the [weekly performance chat](https://make.wordpress.org/core/tag/performance/). For an updated list of POCs, please [see the `CODEOWNERS` file](https://github.com/WordPress/performance/blob/trunk/.github/CODEOWNERS).

Reviewer: I know this is not the intention, but it reads a bit like gate keeping to me (as in "is it okay that I work on this?"). It would be good to clarify that you should notify the POC at any relevant stage so that they're aware. It's best to do that before actually starting the work, but the contributor can e.g. open a PR at any time.

The main reason to notify on the issue would be so that it is not picked up by another contributor at the similar time and there would be conflicting pull requests - that would not be an efficient or fair use of either people's time. So it's also important for any other potential contributor to that issue.

- Change the Project column to **In Progress**
- Remove the **Needs Dev** label, if applicable

## Needs Discussion
Many issues require discussion and definition before development begins. For example, multiple approaches may be considered and the community should discuss them before a contributor proceeds with development. For those issues, add the **Needs Discussion** label and raise the issue in the weekly performance chat.

## Needs Decision
After discussion, a formal vote may be needed to determine how to proceed. If that’s the case, please tag @bethanylang in the issue for assistance with setting up a vote via GitHub comment. [An example vote can be found here](https://github.com/WordPress/performance/issues/92#issuecomment-1068215411).

## Needs Dev
If an issue requires (more) development and you are unable to complete it yourself, remove yourself as the assignee and add a **Needs Dev** label.

> ## Needs Testing

Reviewer: Shouldn't these be `h3` headings (`###`)? I think it doesn't make sense to have them at the same level as "Opening an issue". Maybe we can group them into another higher-level `h2` heading like "Workflow labels", or leave them under the issues section?

If you have completed initial engineering and want community members to test prior to merge, add the Needs Testing label.

> # Pull requests

Reviewer: I think this needs to be an `h2` instead, this way it falls into the same level as "Opening an issue" and "Trac tickets". Also, there should only ever be a single `h1` (which should be the overall document title, see my comment above).

All pull requests must:

- Be associated with an issue
- Have [Type] and [Focus] label matching the related issue’s [Type] and [Focus] labels
- Have either a milestone or the “no milestone” label

When a PR for an issue is ready for review:
- Change the Project column on the issue to **Review**
- Add the **Needs Review** label to the issue

Reviewers will be auto-assigned based upon the issue’s assigned Project. Note that all PRs require at least two reviewers.

## Trac tickets
[Trac](https://core.trac.wordpress.org/) is the system used to track issues related to WordPress core. There is a [‘performance’ label in Trac](https://core.trac.wordpress.org/query?status=!closed&focuses=~performance) that has been added to over 300 issues dating back nearly 15 years. 

While the Performance team’s focus is on the work in the GitHub repo, the community is welcome to consult with the team on performance-related Trac issues. There may also be issues in Trac that are duplicates of planned work in the GitHub repo, or warrant further discussion for inclusion in the plugin.

> Initial discussion on performance-related Trac issues should take place in the Trac issue. When an issue is ready to begin development:

Reviewer: Nit-pick:
```suggestion
Initial discussion on performance-related Trac tickets should take place in the Trac ticket. When an issue is ready to begin development:
```

In GitHub:
- If it is a smaller issue or patch, no need to add to GitHub; manage in Trac
- If it is a larger issue, create a new GitHub issue for it in the Performance repo, following the workflow outlined in this document
- Include a link to the related Trac issue in the issue description

In Trac:
- Add a link to the related GH issue
- Change the status to assigned

> Change the Owner to yourself

Reviewer: I think this section is still a bit preliminary. There's more complexity to this, it really depends on each individual Trac ticket whether it is something worth (or even _possible_) to implement in the Performance Lab plugin, or whether it needs to be implemented against WP core immediately.

I wonder if we should for now keep the "Trac tickets" section limited to the brief introduction above this marked section and define a proper workflow around Trac separately. We can later expand this doc as needed. I think that would allow us to get all the rest of it which has been clear for a while merged quickly.
